modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RecoverFailureDocumentProcessor.java

```diff
@@ -0,0 +1,126 @@
+    public IngestDocument execute(IngestDocument document) throws Exception {
+        if (document.hasField("document") == false) {
+            throw new IllegalArgumentException("field [document] not present as part of path [document]");
+        }
+
+        if (document.hasField("document.source") == false) {
+            throw new IllegalArgumentException("field [source] not present as part of path [document.source]");
+        }
+
+        if (document.hasField("error") == false) {
+            throw new IllegalArgumentException("field [error] not present as part of path [error]");
+        }
+
+        // store pre-recovery data in ingest metadata
+        storePreRecoveryData(document);
+
+        // Get the nested 'document' field, which holds the original document and metadata.
+        Map<String, Object> failedDocument = (Map<String, Object>) document.getFieldValue("document", Map.class);
+
+        // Copy the original index, routing, and id back to the document's metadata.
+        String originalIndex = (String) failedDocument.get("index");
+        if (originalIndex != null) {
+            document.setFieldValue("_index", originalIndex);
+        }
+
+        String originalRouting = (String) failedDocument.get("routing");
+        if (originalRouting != null) {
+            document.setFieldValue("_routing", originalRouting);
+        }
+
+        String originalId = (String) failedDocument.get("id");
+        if (originalId != null) {
+            document.setFieldValue("_id", originalId);
+        }
+
+        // Get the original document's source.
+        Map<String, Object> originalSource = (Map<String, Object>) failedDocument.get("source");
+
+        // Overwrite the _source with original source contents.
+        Map<String, Object> source = document.getSource();
+        source.clear();
+        source.putAll(originalSource);
+
+        // Return the modified document.
+        return document;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    public static final class Factory implements Processor.Factory {
+        @Override
+        public RecoverFailureDocumentProcessor create(
+            Map<String, Processor.Factory> registry,
+            String processorTag,
+            String description,
+            Map<String, Object> config,
+            ProjectId projectId
+        ) throws Exception {
+            return new RecoverFailureDocumentProcessor(processorTag, description);
+        }
+    }
+
+    private static void storePreRecoveryData(IngestDocument document) {
+        Map<String, Object> sourceAndMetadataMap = document.getSourceAndMetadata();
+
+        // Create the pre_recovery data structure
+        Map<String, Object> preRecoveryData = new HashMap<>();
+
+        // Copy everything from the current document
+        sourceAndMetadataMap.forEach((key, value) -> {
+            if ("document".equals(key) && value instanceof Map) {
```

Comment:
We reuse a lot of these failure store field names inline all around the file. Maybe we could define some constants for them at the top of the file to avoid magic strings